madanalysis/layout/histogram.py

```diff
@@ -1,125 +1,144 @@
 ################################################################################
-#  
+#
 #  Copyright (C) 2012-2023 Jack Araz, Eric Conte & Benjamin Fuks
 #  The MadAnalysis development team, email: <ma5team@iphc.cnrs.fr>
-#  
+#
 #  This file is part of MadAnalysis 5.
 #  Official website: <https://github.com/MadAnalysis/madanalysis5>
-#  
+#
 #  MadAnalysis 5 is free software: you can redistribute it and/or modify
 #  it under the terms of the GNU General Public License as published by
 #  the Free Software Foundation, either version 3 of the License, or
 #  (at your option) any later version.
-#  
+#
 #  MadAnalysis 5 is distributed in the hope that it will be useful,
 #  but WITHOUT ANY WARRANTY; without even the implied warranty of
 #  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 #  GNU General Public License for more details.
-#  
+#
 #  You should have received a copy of the GNU General Public License
 #  along with MadAnalysis 5. If not, see <http://www.gnu.org/licenses/>
-#  
+#
 ################################################################################
 
 
 from __future__ import absolute_import
-from madanalysis.layout.histogram_core import HistogramCore
 import logging
-from six.moves import range
+import numpy as np
+from madanalysis.layout.histogram_core import HistogramCore
 
 
 class Histogram:
-
     def __init__(self):
         self.Reset()
 
-
     def Print(self):
 
         # General info
-        inform = self.name + ' ' + str(self.nbins) + str(self.xmin) + ' ' + str(self.xmax)
-        if self.ymin!=[] or self.ymax!=[]:
-           inform = inform + ' ' + str(self.ymin) + ' ' + str(self.ymax)
-        logging.getLogger('MA5').info(inform)
+        inform = self.name + " " + str(self.nbins) + str(self.xmin) + " " + str(self.xmax)
+        if self.ymin != [] or self.ymax != []:
+            inform = inform + " " + str(self.ymin) + " " + str(self.ymax)
+        logging.getLogger("MA5").info(inform)
 
         # Data
         self.positive.Print()
         self.negative.Print()
         self.summary.Print()
 
+    def CreateHistogram(self):
+        pass
+
+    def FinalizeReading(self, main, dataset):
 
-    def FinalizeReading(self,main,dataset):
+        self.positive.convert_to_numpy()
+        self.negative.convert_to_numpy()
+        self.summary.convert_to_numpy()
 
         # Statistics
-        self.summary.nevents   = self.positive.nevents   + self.negative.nevents
-        self.summary.nentries  = self.positive.nentries  + self.negative.nentries
+        self.summary.nevents = self.positive.nevents + self.negative.nevents
+        self.summary.nentries = self.positive.nentries + self.negative.nentries
 
         # sumw
-        self.summary.sumw      = self.positive.sumw      - self.negative.sumw
-        if self.summary.sumw<0:
-            self.summary.sumw=0
+        self.summary.sumw = np.clip(self.positive.sumw - self.negative.sumw, 0, None)
```

Comment:
Mmmmh. This I have to think. A negative bin hence becomes 0. I think that it is better to keep its negative value and plot the absolute value with a different line style.